File: src/com/gmail/hexragonat/clockGadget/controller/MainController.java
package com.gmail.hexragonat.clockGadget.controller;

import com.gmail.hexragonat.clockGadget.ClockApp;
import com.gmail.hexragonat.clockGadget.WordEnum;
import com.jfoenix.controls.JFXDrawer;
import javafx.animation.FillTransition;
import javafx.application.Platform;
import javafx.fxml.FXML;
import javafx.scene.Node;
import javafx.scene.effect.Glow;
import javafx.scene.layout.GridPane;
import javafx.scene.paint.Color;
import javafx.scene.paint.Paint;
import javafx.scene.text.Text;
import javafx.stage.Stage;
import javafx.util.Duration;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.concurrent.TimeUnit;

/**
 * Controller class for the JavaFX program.
 * @see ClockApp
 */
public class MainController
{
	private final ClockApp app;
	private final Stage stage;


	@FXML
	protected GridPane root;
	@FXML
	protected GridPane letterGrid;
	@FXML
	protected Text closeLabel;
	@FXML
	protected Text settingsLabel;
	@FXML
	public JFXDrawer sidebar;


	private String offColor = "#333333";
	private String onColor = "#1a75ff";

	private double xOffset = 0;
	private double yOffset = 0;

	/*
	 * Dependency injection.
	 */
	public MainController(ClockApp app)
	{
		this.app = app;
		this.stage = app.stage;
	}

	public void setOnColor(String onColor)
	{
		this.onColor = onColor;
	}

	public void setOffColor(String offColor)
	{
		this.offColor = offColor;
	}

	public void setBackgroundColor(String color)
	{
		root.setStyle("-fx-background-color: "+color+";");
	}

	public void setup()
	{
		/*
		 * Stage dragging property.
		 */
		root.setOnMousePressed(event ->
		{
			xOffset = event.getSceneX();
			yOffset = event.getSceneY();
		});

		final Stage finalStage = stage;
		root.setOnMouseDragged(event ->
		{
			finalStage.setX(event.getScreenX() - xOffset);
			finalStage.setY(event.getScreenY() - yOffset);
		});

		root.setOnMouseEntered(event -> setClockField(null));
		root.setOnMouseExited(event -> resetClockField());

		/*
		 * Close button handling.
		 */
		closeLabel.setOnMouseEntered(event ->
		{
			FillTransition ft = new FillTransition(Duration.millis(200), closeLabel, Color.valueOf(offColor), Color.valueOf(onColor).brighter());
			ft.play();

			setClockField(" CLOSE ");
		});
		closeLabel.setOnMouseExited(event ->
		{
			FillTransition ft = new FillTransition(Duration.millis(200), closeLabel, Color.valueOf(onColor).brighter(), Color.valueOf(offColor));
			ft.play();

			setClockField(null);
		});
		closeLabel.setOnMouseClicked(event ->
		{
			app.heartbeat.stop();
			stage.close();
			app.settingsManager.save();
			System.exit(0);
		});

		settingsLabel.setOnMouseEntered(event ->
		{
			FillTransition ft = new FillTransition(Duration.millis(200), settingsLabel, Color.valueOf(offColor), Color.valueOf(onColor).brighter());
			ft.play();
			setClockField("OPTIONS");
		});
		settingsLabel.setOnMouseExited(event ->
		{
			FillTransition ft = new FillTransition(Duration.millis(200), settingsLabel, Color.valueOf(onColor).brighter(), Color.valueOf(offColor));
			ft.play();

			if (WordEnum.P_OCLOCK.isActive())
			{
				FillTransition ft0 = new FillTransition(Duration.millis(200), settingsLabel, Color.valueOf(offColor), Color.valueOf(onColor));
				ft0.play();
			}

			setClockField(null);
		});
		settingsLabel.setOnMouseClicked(event ->
		{
			System.out.println(sidebar.isHidden());
			if (sidebar.isHidden()) sidebar.open();
			else sidebar.close();
		});


		resetAll();
	}

	public void setClockField(String string)
	{
		if (string == null || string.isEmpty())
		{
			SimpleDateFormat formatter = new SimpleDateFormat("hh:mma");
			string = formatter.format(new Date());
		}

		for (int i = 0; i < 7; i++)
		{
			final int finalI = i;
			final String finalString = string;
			app.heartbeat.exec.schedule(() ->
					Platform.runLater(() ->
					{
						Text node = letterAt(finalI +2,5);

						if (node != null)
						{
							node.setStyle("-fx-font-weight: bold");
							node.setFill(Paint.valueOf(Color.valueOf(onColor).brighter().toString()));
							node.setText(String.valueOf(finalString.charAt(finalI)));
						}
					}), 50 * (i+1), TimeUnit.MILLISECONDS);
		}
	}

	public void resetClockField()
	{
		String original = "IRTHWDA";

		for (int i = 0; i < 7; i++)
		{
			final int finalI = i;
			app.heartbeat.exec.schedule(() ->
					Platform.runLater(() ->
					{
						Text node = letterAt(finalI+2,5);

						if (node != null)
						{
							node.setStyle(null);
							node.setFill(Paint.valueOf(offColor));
							node.setText(String.valueOf(original.charAt(finalI)));
						}
					}), 50 * (i+1), TimeUnit.MILLISECONDS);
		}
	}

	/**
	 * Grab a letter from the letter grid.
	 */
	private Text letterAt(Integer x, Integer y)
	{
		for (Node node : letterGrid.getChildren())
		{
			Integer columnIndex = GridPane.getColumnIndex(node);
			if (columnIndex == null) columnIndex = 0;

			Integer rowIndex = GridPane.getRowIndex(node);
			if (rowIndex == null) rowIndex = 0;

			if (columnIndex.equals(x) && rowIndex.equals(y))
			{
				return ((Text) node);
			}
		}
		return null;
	}

	/**
	 * Toggle the clock's letters by the words.
	 * @see WordEnum
	 */
	public void toggle(WordEnum wordEnum, boolean bool)
	{
		if (wordEnum != null)
		{
			try
			{
				int i = 0;
				int max = wordEnum.size;

				/*
				 * Overlapping words.
				 */
				if (wordEnum == WordEnum.M_TWENTYFIVE && (WordEnum.M_TWENTY.isActive() || WordEnum.M_FIVE.isActive()))
				{
					if (WordEnum.M_TWENTY.isActive()) i = 6;
					else if (WordEnum.M_FIVE.isActive()) max = 6;
				}
				else if ((wordEnum == WordEnum.M_TWENTY || wordEnum == WordEnum.M_FIVE) && WordEnum.M_TWENTYFIVE.isActive())
				{
					return;
				}

				for (; i < max; i++)
				{
					Text node = letterAt(wordEnum.xOf[i], wordEnum.yOf[i]);

					if (node != null)
					{
						if (!bool)
						{
							if (wordEnum.isActive())
							{
								FillTransition ft = new FillTransition(Duration.millis(1000), node, Color.valueOf(onColor), Color.valueOf(offColor));
								ft.play();
							}

							node.setFill(Paint.valueOf(offColor));
							node.setEffect(null);
						}
						else
						{
							if (!wordEnum.isActive())
							{
								FillTransition ft = new FillTransition(Duration.millis(1000), node, Color.valueOf(offColor), Color.valueOf(onColor));
								ft.play();
							}

							node.setFill(Paint.valueOf(onColor));
							node.setEffect(new Glow(5.0));
						}
					}
				}
				wordEnum.setActive(bool);
			}
			catch (NullPointerException e)
			{
				e.printStackTrace();
			}
		}
	}

	/*
	 * Turn all letters off.
	 */
	protected void clearAll()
	{
		for (WordEnum enum0 : WordEnum.values())
		{
			toggle(enum0, false);
		}
	}

	protected void resetAll()
	{
		// set all of letters color to off color
		for (int x = 0; x <= 10; x++)
		{
			for (int y = 0; y <= 10; y++)
			{
				Text node = letterAt(x, y);
				if (node != null)
				{
					node.setFill(Paint.valueOf(offColor));
					node.setEffect(null);
				}
			}
		}
	}

}

File: src/com/gmail/hexragonat/clockGadget/controller/MainController.kt
package com.gmail.hexragonat.clockGadget.controller

import com.gmail.hexragonat.clockGadget.ClockApp
import com.gmail.hexragonat.clockGadget.WordEnum
import com.jfoenix.controls.JFXDrawer
import javafx.animation.FillTransition
import javafx.application.Platform
import javafx.fxml.FXML
import javafx.scene.effect.Glow
import javafx.scene.layout.GridPane
import javafx.scene.paint.Color
import javafx.scene.paint.Paint
import javafx.scene.text.Text
import javafx.stage.Stage
import javafx.util.Duration
import java.text.SimpleDateFormat
import java.util.Date
import java.util.concurrent.TimeUnit

/**
 * Controller class for the JavaFX program.
 * @see ClockApp
 */
class MainController
(private val app : ClockApp)
{
    private val stage : Stage
    
    @FXML @JvmField
    var root : GridPane? = null
    @FXML @JvmField
    var letterGrid : GridPane? = null
    @FXML @JvmField
    var closeLabel : Text? = null
    @FXML @JvmField
    var settingsLabel : Text? = null
    @FXML @JvmField
    var sidebar : JFXDrawer? = null
    
    private var offColor = "#333333"
    private var onColor = "#1a75ff"
    
    private var xOffset = 0.0
    private var yOffset = 0.0
    
    init
    {
        this.stage = app.stage
    }
    
    fun setOnColor(onColor : String)
    {
        this.onColor = onColor
    }
    
    fun setOffColor(offColor : String)
    {
        this.offColor = offColor
    }
    
    fun setBackgroundColor(color : String)
    {
        root!!.style = "-fx-background-color: $color;"
    }
    
    fun setup()
    {
        /*
		 * Stage dragging property.
		 */
        root!!.setOnMousePressed { event ->
            xOffset = event.sceneX
            yOffset = event.sceneY
        }
        
        val finalStage = stage
        root!!.setOnMouseDragged { event ->
            finalStage.x = event.screenX - xOffset
            finalStage.y = event.screenY - yOffset
        }
        
        root!!.setOnMouseEntered { event -> setClockField(null) }
        root!!.setOnMouseExited { event -> resetClockField() }
        
        setupCloseLabel()
        setupSettingsLabel()
    }
    
    private fun setupCloseLabel()
    {
        /*
		 * Close button handling.
		 */
        closeLabel!!.setOnMouseEntered { event ->
            val ft = FillTransition(Duration.millis(200.0), closeLabel, Color.valueOf(offColor), Color.valueOf(onColor).brighter())
            ft.play()
            closeLabel!!.effect = Glow(5.0)
            setClockField(" CLOSE ")
        }
        closeLabel!!.setOnMouseExited { event ->
            val ft = FillTransition(Duration.millis(200.0), closeLabel, Color.valueOf(onColor).brighter(), Color.valueOf(offColor))
            ft.play()
            closeLabel!!.effect = null
            setClockField(null)
        }
        closeLabel!!.setOnMouseClicked { event ->
            app.heartbeat.stop()
            stage.close()
            app.settingsManager.save()
            System.exit(0)
        }
    }
    
    private fun setupSettingsLabel()
    {
        settingsLabel!!.setOnMouseEntered { event ->
            val ft = FillTransition(Duration.millis(200.0), settingsLabel, Color.valueOf(offColor), Color.valueOf(onColor).brighter())
            ft.play()
            settingsLabel!!.effect = Glow(5.0)
            setClockField("OPTIONS")
        }
        settingsLabel!!.setOnMouseExited { event ->
            val ft = FillTransition(Duration.millis(200.0), settingsLabel, Color.valueOf(onColor).brighter(), Color.valueOf(offColor))
            ft.play()
            
            if (WordEnum.P_OCLOCK.isActive)
            {
                val ft0 = FillTransition(Duration.millis(200.0), settingsLabel, Color.valueOf(offColor), Color.valueOf(onColor))
                ft0.play()
            }
            else
            {
                settingsLabel!!.effect = null
            }
            
            setClockField(null)
        }
        settingsLabel!!.setOnMouseClicked { event ->
            if (sidebar!!.isHidden())
                sidebar!!.open()
            else
                sidebar!!.close()
        }
    }
    
    fun setClockField(string : String?)
    {
        var string0 = string
        if (string0 == null || string0.isEmpty())
        {
            val formatter = SimpleDateFormat("hh:mma")
            string0 = formatter.format(Date())
        }
        
        for (i in 0 .. 6)
        {
            val finalI = i
            val finalString = string0
            app.heartbeat.executor.schedule({
                Platform.runLater {
                    val node = letterAt(finalI + 2, 5)
                    
                    if (node != null)
                    {
                        node.style = "-fx-font-weight: bold"
                        node.fill = Paint.valueOf(Color.valueOf(onColor).brighter().toString())
                        node.text = finalString!![finalI].toString()
                        node.effect = Glow(1.0)
                    }
                }
            }, (50 * (i + 1)).toLong(), TimeUnit.MILLISECONDS)
        }
    }
    
    fun resetClockField()
    {
        val original = "EXAVALO"
        
        for (i in 0 .. 6)
        {
            val finalI = i
            app.heartbeat.executor.schedule({
                Platform.runLater {
                    val node = letterAt(finalI + 2, 5)
                    
                    if (node != null)
                    {
                        node.style = null
                        node.fill = Paint.valueOf(offColor)
                        node.text = original[finalI].toString()
                        node.effect = null
                    }
                }
            }, (50 * (i + 1)).toLong(), TimeUnit.MILLISECONDS)
        }
    }
    
    /**
     * Grab a letter from the letter grid.
     */
    private fun letterAt(x : Int?, y : Int?) : Text?
    {
        for (node in letterGrid!!.children)
        {
            var columnIndex = GridPane.getColumnIndex(node)
            if (columnIndex == null) columnIndex = 0
            
            var rowIndex = GridPane.getRowIndex(node)
            if (rowIndex == null) rowIndex = 0
            
            if (columnIndex == x && rowIndex == y)
            {
                return node as Text
            }
        }
        return null
    }
    
    /**
     * Toggle the clock's letters by the words.
     * @see WordEnum
     */
    fun toggle(wordEnum : WordEnum?, bool : Boolean)
    {
        if (wordEnum != null)
        {
            try
            {
                var i = 0
                var max = wordEnum.size
                
                /*
				 * Overlapping words.
				 */
                if (wordEnum == WordEnum.M_TWENTYFIVE && (WordEnum.M_TWENTY.isActive || WordEnum.M_FIVE.isActive))
                {
                    if (WordEnum.M_TWENTY.isActive)
                        i = 6
                    else if (WordEnum.M_FIVE.isActive) max = 6
                }
                else if ((wordEnum == WordEnum.M_TWENTY || wordEnum == WordEnum.M_FIVE) && WordEnum.M_TWENTYFIVE.isActive)
                {
                    return
                }
                
                while (i < max)
                {
                    val node = letterAt(wordEnum.xOf[i], wordEnum.yOf[i])
                    
                    if (node != null)
                    {
                        if (!bool)
                        {
                            if (wordEnum.isActive)
                            {
                                val ft = FillTransition(Duration.millis(1000.0), node, Color.valueOf(onColor), Color.valueOf(offColor))
                                ft.play()
                            }
                            
                            node.fill = Paint.valueOf(offColor)
                            node.effect = null
                        }
                        else
                        {
                            if (!wordEnum.isActive)
                            {
                                val ft = FillTransition(Duration.millis(1000.0), node, Color.valueOf(offColor), Color.valueOf(onColor))
                                ft.play()
                            }
                            
                            node.fill = Paint.valueOf(onColor)
                            node.effect = Glow(5.0)
                        }
                    }
                    i++
                }
                wordEnum.isActive = bool
            }
            catch (e : NullPointerException)
            {
                e.printStackTrace()
            }
            
        }
    }
    
    fun getSidebar() = sidebar
    
    /*
	 * Turn all letters off.
	 */
    fun clearAll()
    {
        for (enum0 in WordEnum.values())
        {
            toggle(enum0, false)
        }
    }
    
    fun resetAll()
    {
        // set all of letters color to off color
        for (x in 0 .. 10)
        {
            for (y in 0 .. 10)
            {
                val node = letterAt(x, y)
                if (node != null)
                {
                    node.fill = Paint.valueOf(offColor)
                    node.effect = null
                }
            }
        }
    }
}
